spec/models/html_document_spec.rb

```diff
@@ -270,9 +270,142 @@
+    end
+
+    context 'when a dcterms subject is available' do
+      let(:raw_document) do
+        <<~HTML
+          <html lang="en">
+            <head>
+              <title>My Title</title>
+              <meta name="dcterms.subject" content="One DCTerms Subject, Another DCTerms Subject"/>
+            </head>
+          </html>
+        HTML
+      end
+
+      it { is_expected.to eq 'One DCTerms Subject, Another DCTerms Subject' }
+    end
+
+    context 'when a dcterms keywords is available' do
+      let(:raw_document) do
+        <<~HTML
+          <html lang="en">
+            <head>
+              <title>My Title</title>
+              <meta name="dcterms.keywords" content="One DCTerms Keyword, Another DCTerms Keyword"/>
+            </head>
+          </html>
+        HTML
+      end
+
+      it { is_expected.to eq 'One DCTerms Keyword, Another DCTerms Keyword' }
+    end
+
+    context 'when all dc keyword sources are available' do
       let(:raw_document) { doc_with_dc_data }
 
-      it { is_expected.to eq 'One Subject, Another Subject' }
+      it { is_expected.to include('One DC Subject') }
+      it { is_expected.to include('One DCTerms Subject') }
+      it { is_expected.to include('One DCTerms Keyword') }
+      it { is_expected.to include('Another DC Subject') }
+      it { is_expected.to include('Another DCTerms Subject') }
+      it { is_expected.to include('Another DCTerms Keyword') }
+    end
+
+    context 'when a meta keywords is available' do
+      it { is_expected.to eq 'this, that' }
+    end
+
+    context 'when parallel article:tags are available' do
+      let(:raw_document) do
+        <<~HTML
+          <html lang="en">
+            <head>
+              <title>My Title</title>
+              <meta property="article:tag" content="article tag 1" />
+              <meta property="article:tag" content="article tag 2" />
+              <meta property="article:tag" content="article tag 3" />
+            </head>
+          </html>
+        HTML
+      end
+
+      it { is_expected.to eq 'article tag 1, article tag 2, article tag 3' }
+    end
+
+    context 'when an article:section is available' do
+      let(:raw_document) do
+        <<~HTML
+          <html lang="en">
+            <head>
+              <title>My Title</title>
+              <meta property="article:section" content="article section">
+            </head>
+          </html>
+        HTML
+      end
+
+      it { is_expected.to eq 'article section' }
+    end
+
+    context 'when meta keywords and article keywords are available' do
+      let(:raw_document) { read_fixture_file('/html/page_with_og_metadata.html') }
+
+      it { is_expected.to include('this') }
+      it { is_expected.to include('that') }
+      it { is_expected.to include('the other') }
+      it { is_expected.to include('thing') }
+    end
+
+    context 'when parallel keywords duplicate a list of keywords' do
```

Comment:
Are there any scenarios where we would need to dedupe keywords that aren't perfectly replicated in both a comma-separated list and a set of separate tags? 

I believe something like the following (only two tags in the comma-separated list) would fail, but not sure if it's a likely scenario:

```html
<meta name="keywords" content="tag 1, tag 3">
<meta property="article:section" content="tag 1" />
<meta property="article:section" content="tag 2" />
<meta property="article:section" content="tag 3" />
```

Reply:
Yes, you're correct, that would keep the duplicated "tag 3" around:

```
[1] pry(main)> doc =         <<~HTML
          <html lang="en">
            <head>
              <title>My Title</title>
              <meta name="keywords" content="tag 1, tag 3">
              <meta property="article:section" content="tag 1" />
              <meta property="article:section" content="tag 2" />
              <meta property="article:section" content="tag 3" />
            </head>
          </html>
        HTML
=> "<html lang=\"en\">\n  <head>\n    <title>My Title</title>\n    <meta name=\"keywords\" content=\"tag 1, tag 3\">\n    <meta property=\"article:section\" content=\"tag 1\" />\n    <meta property=\"article:section\" content=\"tag 2\" />\n    <meta property=\"article:section\" content=\"tag 3\" />\n  </head>\n</html>\n"
[2] pry(main)> doc = HtmlDocument.new(document: doc, url:'www.example.com')
=> #<HtmlDocument:0x00007fe3d72263f8
 @document=
  "<html lang=\"en\">\n  <head>\n    <title>My Title</title>\n    <meta name=\"keywords\" content=\"tag 1, tag 3\">\n    <meta property=\"article:section\" content=\"tag 1\" />\n    <meta property=\"article:section\" content=\"tag 2\" />\n    <meta property=\"article:section\" content=\"tag 3\" />\n  </head>\n</html>\n",
 @url="www.example.com">
[3] pry(main)> doc.send(:extract_keywords)
=> [nil, nil, nil, "tag 1, tag 3", nil, "tag 1, tag 2, tag 3"]
[4] pry(main)> doc.keywords
=> "tag 1, tag 3, tag 1, tag 2, tag 3"
```

I'm inclined to keep this as is, but welcome feedback.  First, because I agree that we don't know how often (or if ever) we will encounter this.  Across my list of test sites the most common pattern I saw was exact duplication of content across one or more metadata elements (aka https://www.niams.nih.gov/ where meta keywords directly duplicate parallel article:tag elements).  I think this is probably a safe assumption given the wide adoption of content management systems, but it is definitely an _assumption_.  But, also, taking the opposite approach - treating every word offset by commas as a truly unique entity to be deduped - runs the risk of improperly deleting semantically different content (e.g. a list of one element containing "Smith, John" and another containing "Smythe, John" would drop one of the Johns).  Whereas, the risk with the current approach is merely improperly duplicating things (but not deleting things).  

But, this conversation is still helpful because it does raise to my awareness another future pitfall with maintaining the current system of generating tags as a comma-separated list.  As we get deeper into the tags indexing and, more importantly, SERP design, we'll need to be aware that phrases like my example above "Smith, John" are, unfortunately, going to at this particular moment be treated as two different keywords.  Or maybe more likely for the type of content we see, geographic tags like "Washington, D.C." or "Baltimore, MD").

Reply:
Oooh, these are very interesting scenarios! In that case, I'm inclined to agree that unintentional deduping seems worse than leaving some duplicate tags in. There also doesn't seem to be a risk, other than requiring infinitesimally more storage, to retaining a couple of duplicate tags for situations like these.